app/services/onlinetesting/Phase2TestService.scala

```diff
@@ -250,17 +250,24 @@ trait Phase2TestService extends OnlineTestService with ScheduleSelector {
     }
   }
 
-  def markAsCompleted(cubiksUserId: Int)(implicit hc: HeaderCarrier, rh: RequestHeader): Future[Unit] = eventSink {
-    updatePhase2Test(cubiksUserId, phase2TestRepo.updateTestCompletionTime(_:Int, dateTimeFactory.nowLocalTimeZone)).flatMap { u =>
-      require(u.phase2TestGroup.activeTests.nonEmpty, "Active tests cannot be found")
-      val activeTestsCompleted = u.phase2TestGroup.activeTests forall (_.completedDateTime.isDefined)
-      activeTestsCompleted match {
-        case true =>
-          phase2TestRepo.updateProgressStatus(u.applicationId, ProgressStatuses.PHASE2_TESTS_COMPLETED) map { _ =>
+  private def isTestGroupExpired(cubiksUserId: Int) = for {
+    profile <- phase2TestRepo.getTestProfileByCubiksId(cubiksUserId)
+  } yield profile.phase2TestGroup.expirationDate.isBeforeNow()
+
+  def markAsCompleted(cubiksUserId: Int)(implicit hc: HeaderCarrier, rh: RequestHeader): Future[Unit] = isTestGroupExpired(cubiksUserId).flatMap {
```

Comment:
It seems like doing the query to phase2TestRepo.getTestProfileByCubiksId(cubiksUserId) in the isTestGroupExpired method is introducing another call, and perhaps a race condition. Is it possible to put a query in the updatePhase2Test that only performs the update if necessary